tests/core/builder-tools/test_chain_builder.py

```diff
@@ -0,0 +1,160 @@
+import pytest
+
+from cytoolz import pipe
+
+from eth_utils import ValidationError
+
+from eth.chains.base import (
+    Chain,
+    MiningChain,
+)
+from eth.tools.builder.chain import (
+    frontier_at,
+    genesis,
+    mine_block,
+    mine_blocks,
+    disable_pow_check,
+    fork_chain,
+    at_block_number,
+)
+
+
+MINING_CHAIN_PARAMS = (
+    MiningChain,
+    frontier_at(0),
+    disable_pow_check,
+    genesis(),
+)
+
+
+@pytest.fixture
+def mining_chain():
+    return pipe(*MINING_CHAIN_PARAMS)
+
+
+REGULAR_CHAIN_PARAMS = (
+    Chain,
+    frontier_at(0),
+    genesis(),
+)
+
+
+@pytest.fixture
+def regular_chain():
+    return pipe(*REGULAR_CHAIN_PARAMS)
+
+
+@pytest.fixture(params=(MINING_CHAIN_PARAMS, REGULAR_CHAIN_PARAMS))
+def any_chain(request):
+    return pipe(*request.param)
+
+
+def test_chain_builder_build_single_default_block(mining_chain):
+    chain = pipe(
+        mining_chain,
+        mine_block(),
+    )
+
+    header = chain.get_canonical_head()
+    assert header.block_number == 1
+
+
+def test_chain_builder_build_two_default_blocks(mining_chain):
+    chain = pipe(
+        mining_chain,
+        mine_block(),
+        mine_block(),
+    )
+
+    header = chain.get_canonical_head()
+    assert header.block_number == 2
+
+
+def test_chain_builder_build_mine_multiple_blocks(mining_chain):
+    chain = pipe(
+        mining_chain,
+        mine_blocks(5),
+    )
+
+    header = chain.get_canonical_head()
+    assert header.block_number == 5
+
+
+def test_chain_builder_mine_block_with_parameters(mining_chain):
+    chain = pipe(
+        mining_chain,
+        mine_block(extra_data=b'test-setting-extra-data'),
+    )
+
+    header = chain.get_canonical_head()
+    assert header.extra_data == b'test-setting-extra-data'
+
+
+def test_chain_builder_mine_block_only_on_mining_chain(regular_chain):
+    with pytest.raises(ValidationError, match="MiningChain"):
+        mine_block()(regular_chain)
+
+
+def test_chain_builder_mine_multiple_blocks_only_on_mining_chain(regular_chain):
+    with pytest.raises(ValidationError, match="MiningChain"):
+        mine_blocks(5)(regular_chain)
+
+
+def test_chain_builder_fork_chain(mining_chain):
+    pre_fork_chain = pipe(
+        mining_chain,
+        mine_block(),  # 1
+        mine_block(),  # 2
+        mine_block(),  # 3
+    )
+    block_1 = pre_fork_chain.get_canonical_block_by_number(1)
+    block_2 = pre_fork_chain.get_canonical_block_by_number(2)
+    block_3 = pre_fork_chain.get_canonical_block_by_number(3)
+
+    chain = pipe(
+        pre_fork_chain,
+        fork_chain(
+            at_block_number(2),
+            mine_block(extra_data=b'fork-it!'),  # fork 3
+            mine_block(),  # fork 4
+            mine_block(),  # fork 5
+        ),
+    )
```

Comment:
So, this wasn't intentional but looking at this I think I understand why this way of expressing what should happen in the chain is so much easier to read.

The indentations provide a level of structure that literally gives it physical shape.  I haven't done transactions yet, but I'm thinking about this API.

```python
mine_block(
    txn(sender=account_0),
    txn(senter=account_1, to='0x0123...')
    txn(), # safe defaults
    extra_data=b'heres-some-extra-data',
)
```

It visually contains the transactions within the block.  Above the same thing with the `fork_chain`

Reply:
For forks, I want to do something like this to try and visually group the forks.

```python
chain = pipe(
    chain,
    mine_block(),
    chain_split(
        (mine_block(), mine_block(), mine_block()),  # fork A
        (mine_block(extra_data=b'...'), mine_block(), mine_block(), mine_block(), # fork B
        # results is a multidimensional array with 
        # the intermediate outputs for each fork
        converge_fn=lambda results: results[1][-1],
    )
    mine_block(), # resumes being *single_threaded*,
)
```

I don't love it but it's a starting point.

Reply:
Alright, here's a slightly alternate idea:

```python
manage_chain_a = pack(
    mine_block(),
    mine_block(
        tx(),
        tx(),
    ),
)

manage_chain_b = pack(
    mine_block(extra_data=b'...'),
    mine_block(),
    mine_block(),
)


def merge_chains(results):
    chain_a_results, chain_b_results = zip(*results)

    # do the import


def test_chain_split(chain):
    chain = pipe(
        chain,
        mine_block(),
        mine_block(),
        chain_split(
            manage_chain_a,
            manage_chain_b,
            exit_fn=merge_chains,
        )
        mine_block(tx(), tx(), tx()),
    )
```

The `chain_split` function takes `*args` and each of those can be thought of as a thread (but not an actual thread).  For each function provided as a positional argument call it with an identical copy of the chain.  We call each in succession and then pass all of the results in as a multidimensional array to the `exit_fn`.

Something I still want to figure out is cross communication between the functions.

Reply:
Hm, I think I like the idea of `chain_split` but haven't fallen in :heart: with any of these APIs. Maybe `chain_split` should actually return a tuple of chains (and so is always intended to be used last in the `pipe`):

```py
chaina, chainb = pipe(
    chain,
    mine_block(),
    mine_block(),
    chain_split(
        (mine_block(), ),  # fork A
        (mine_block(extra_data=b'fork-it!'), mine_block(), mine_block(), # fork B
    ),
)
```

---

FWIW, I think the current approach for forking is okay (explicit and clear). Maybe a slightly special version of `pipe` should always be used that never mutates the original chain. Then we wouldn't need a `fork_chain` at all, it would just be:

```py
    pre_fork_chain = build_chain(
        mining_chain,
        mine_block(),  # 1
        mine_block(),  # 2
        mine_block(),  # 3
    )

    fork_chain = build_chain(  # every build_chain automatically forks so the original isn't mutated
        pre_fork_chain,
        at_block_number(2),
        mine_block(extra_data=b'fork-it!'),  # fork 3
        mine_block(),  # fork 4
        mine_block(),  # fork 5
    )
```

I guess the above `chain_split` is easier to read.